### ISSUE TITLE
IRI converted with STR is not comparable with another literal

### ISSUE THREAD

**Issue by vtermanis** [oxigraph/oxigraph#645]
**Describe the bug**
IRI converted with `STR` is not comparable to a literal of the same value

**To Reproduce**
1. Which version of Oxigraph are you using? On which platform?
    - Official `oxigraph/oxigraph:v0.3.19` docker image
3. A command line or a code snippet that triggers the bug.
    - Execute the following query
       ```sparql
        PREFIX p_prefix1: <some:ns>
        PREFIX p_prefix2: <some:ns>

      SELECT * {
          # All three BINDs have literal type in response
          BIND(STR(p_prefix1:) AS ?x)
          BIND(STR(p_prefix2:) AS ?y)
          BIND('some:ns' AS ?z)
          # True (expected)
          FILTER(?x = ?y)
          # Both of these are unexpectedly False
          FILTER(?z = ?z)
          FILTER(?x = 'some:ns')
      }
      ```
    - The above works fine in a few other triple stores I've tried

**Comment by Tpt** [comment]
Hi! Thank you for this great bug report! I have understood the bug: Oxigraph encodes all IRIs using a dictionnary and reuses the same dictionnary for long strings but short strings are inlined. The `STR` function implementation just casts the IRI dictionnary id to a string dictionnary id so in the `FILTER(?x = 'some:ns')` we end up with a comparision between a dictionnary id (left) and an inline string (right) leading to `false` even if they are actually the same string.

Some possible fix:
1. allow inlined IRI when the IRI is small but it will break the on-disk encoding and would require a migration script
2. when casting an IRI to a string, check if the string is small (it will require a read, hence a performance penalty but it's very likely the string will be read anyway at some point when processing the same query result so it might not introduce new I/O reads in a lot of cases).

I would tend to prefer solution 2 for now and plan to do 1 in a future release that will change the on-disk formart.

**Comment by vtermanis** [comment]
Thanks, I tried `ghcr.io/oxigraph/oxigraph:main` docker image and it's working :+1:

**Comment by Tpt** [comment]
@vtermanis Amazing! Thank you!

**Comment by Tpt** [comment]
Released in 0.3.20. Thank you again!
